[ONNX] Deprecated the example_outputs param from torch.onnx.export() function. (#62815)

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64382 [ONNX] Deprecate use_external_data_format param from torch.onnx.export() function. (#62257)
* **#64380 [ONNX] Deprecated the example_outputs param from torch.onnx.export() function. (#62815)**
* #64373 [ONNX] minor doc improvements and cleanup (#62514)
* #64372 [ONNX] Add supplementary tests and description for custom_opsets param from torch.onnx.export() function. (#62085)
* #64371 [ONNX] Remove strip_doc_string param from torch.onnx.export() function. (#61712)
* #64370 [ONNX] Remove argument _retain_param_name from torch.onnx.export() function. (#61702)

* `example_outputs` used to determine the type and shape of the outputs without tracing the execution of the model. And it must be provided when exporting a ScriptModule or ScriptFunction when using export() function.

* Since we can work out `example_outputs` in internal function instead of being provided by user, so we deprecated this argument in the export() function to increase user experience of calling this function.

Co-authored-by: hwangdeyu <dejack953@outlook.com>

Differential Revision: [D30905266](https://our.internmc.facebook.com/intern/diff/D30905266)

hwangdeyu committed:
[ONNX] Deprecated the example_outputs param from torch.onnx.export() function. (#62815)

* `example_outputs` used to determine the type and shape of the outputs without tracing the execution of the model. And it must be provided when exporting a ScriptModule or ScriptFunction when using export() function.

* Since we can work out `example_outputs` in internal function instead of being provided by user, so we deprecated this argument in the export() function to increase user experience of calling this function.

Co-authored-by: hwangdeyu <dejack953@outlook.com>

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Deprecated the example_outputs param from torch.onnx.export() function. (#62815)"


* `example_outputs` used to determine the type and shape of the outputs without tracing the execution of the model. And it must be provided when exporting a ScriptModule or ScriptFunction when using export() function.

* Since we can work out `example_outputs` in internal function instead of being provided by user, so we deprecated this argument in the export() function to increase user experience of calling this function.

Co-authored-by: hwangdeyu <dejack953@outlook.com>

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Deprecated the example_outputs param from torch.onnx.export() function. (#62815)"


* `example_outputs` used to determine the type and shape of the outputs without tracing the execution of the model. And it must be provided when exporting a ScriptModule or ScriptFunction when using export() function.

* Since we can work out `example_outputs` in internal function instead of being provided by user, so we deprecated this argument in the export() function to increase user experience of calling this function.

Co-authored-by: hwangdeyu <dejack953@outlook.com>

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Deprecated the example_outputs param from torch.onnx.export() function. (#62815)"


* `example_outputs` used to determine the type and shape of the outputs without tracing the execution of the model. And it must be provided when exporting a ScriptModule or ScriptFunction when using export() function.

* Since we can work out `example_outputs` in internal function instead of being provided by user, so we deprecated this argument in the export() function to increase user experience of calling this function.

Co-authored-by: hwangdeyu <dejack953@outlook.com>

[ghstack-poisoned]

msaroufim commented:
@msaroufim has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30905266).

BowenBao committed:
Update on "[ONNX] Deprecated the example_outputs param from torch.onnx.export() function. (#62815)"


* `example_outputs` used to determine the type and shape of the outputs without tracing the execution of the model. And it must be provided when exporting a ScriptModule or ScriptFunction when using export() function.

* Since we can work out `example_outputs` in internal function instead of being provided by user, so we deprecated this argument in the export() function to increase user experience of calling this function.

Co-authored-by: hwangdeyu <dejack953@outlook.com>

Differential Revision: [D30905266](https://our.internmc.facebook.com/intern/diff/D30905266)

[ghstack-poisoned]

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -480,13 +493,22 @@ def _model_to_graph(model, args, verbose=False,
+        if example_outputs is None:
+            example_outputs = _get_example_outputs(model, args)
+        else:
+            # example_outpus specified
```
@hwangdeyu updated code to replicate previous results.

hwangdeyu replied:
Looks great to me!

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30905266).

BowenBao committed:
Update on "[ONNX] Deprecated the example_outputs param from torch.onnx.export() function. (#62815)"


* `example_outputs` used to determine the type and shape of the outputs without tracing the execution of the model. And it must be provided when exporting a ScriptModule or ScriptFunction when using export() function.

* Since we can work out `example_outputs` in internal function instead of being provided by user, so we deprecated this argument in the export() function to increase user experience of calling this function.

Co-authored-by: hwangdeyu <dejack953@outlook.com>

Differential Revision: [D30905266](https://our.internmc.facebook.com/intern/diff/D30905266)

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30905266).

Ark-kun commented:
The documentation still mentions the parameter. https://github.com/pytorch/pytorch/blob/9257de7/docs/source/onnx.rst#id30
